test/nn/test_dropout.py

```diff
@@ -0,0 +1,280 @@
+            if p != 0:  # Zero will keep strides as is based on input.
+                # When prob == 0, input stride (54, 18, 6, 2) -> output stride (54, 18, 6, 2)
+                # When prob != 0, input stride (54, 18, 6, 2) -> output stride (27, 9, 3, 1)
+                self.assertTrue(out.is_contiguous(memory_format=memory_format))
+            self.assertEqual(inp_discontiguous, out)
+
+    def _test_dropout_stride_mean_preserve(self, cls, device):
+        def invert_perm(p):
+            d = {x: i for i, x in enumerate(p)}
+            return (d[0], d[1], d[2], d[3])
+
+        inp = torch.ones(2, 3, 4, 5, device=device)
+        shifts = [(0, 0), (1, 0), (0, 1), (1, 1)]
+        for perm in itertools.permutations((0, 1, 2, 3), r=4):
+            for shift in shifts:
+                for p in [1e-10, 0.3, 0.5, 0.7]:
+                    mod = cls(p=p)
+                    permuted_inp = inp.permute(perm).contiguous().permute(invert_perm(perm))
+                    permuted_inp = permuted_inp[shift[0]:, shift[1]:, :, :]
+                    out = mod(permuted_inp)
+
+                    self.assertTrue(out.permute(perm).is_contiguous())
+                    self.assertEqual(inp.mean(), out.mean(), rtol=0.5, atol=0.5)
+                    if p == 1e-10:
+                        self.assertEqual(permuted_inp, out)
+                    else:
+                        self.assertNotEqual(permuted_inp, out)
+
+    def test_Dropout(self, device):
+        input = torch.empty(1000)
+        self._test_dropout(nn.Dropout, device, input)
+
+        self._test_dropout_discontiguous(nn.Dropout, device)
+        self._test_dropout_discontiguous(nn.Dropout, device, memory_format=torch.channels_last)
+
+        self._test_dropout_stride_mean_preserve(nn.Dropout, device)
+
+        if self.device_type == 'cuda' or self.device_type == 'cpu':
+            input = input.bfloat16()
+            self._test_dropout(nn.Dropout, device, input)
+
+    def _test_dropoutNd_no_batch(self, dropout, input):
+        input_clone = input.clone()
+        with freeze_rng_state():
+            res_no_batch = dropout(input)
+
+        with freeze_rng_state():
+            res_batched = dropout(input_clone.unsqueeze(0)).squeeze(0)
+
+        self.assertEqual(res_no_batch, res_batched)
+
+    def _test_dropoutNd_channel_zero(self, dropout, input):
+        # Verify the number of zeros in a channel is 0 or the number of elements in the channel
+        # for a fully positive input tensor
+        shape = input.shape
+        B = shape[0]
+        C = shape[1]
+        channel_numel = torch.tensor(shape[2:]).prod()
+        result = dropout(input)
+
+        for b, c in product(range(B), range(C)):
+            self.assertTrue(result[b, c].count_nonzero() in (0, channel_numel))
+
+    @expectedFailureXLA  # seems like freeze_rng_state is not honoured by XLA
+    def test_Dropout1d(self, device):
+        with set_default_dtype(torch.double):
```

Comment:
Somehow without this, the test fails.

Reply:
Looks like the test is flaky : https://github.com/pytorch/pytorch/issues/84186

Reply:
could you tkae a look at why?

Reply:
Sure.

Reply:
Flaky test failure is related to `dynamo` which I am not sure about.

```
2022-08-28T18:09:40.0858999Z   test_Dropout1d_cpu (__main__.TestNNDeviceTypeCPU) ... torchdynamo.convert_frame: [ERROR] WON'T CONVERT test_Dropout1d test_nn.py line 13722 
2022-08-28T18:09:40.0859583Z due to: 
2022-08-28T18:09:40.0859897Z Traceback (most recent call last):
2022-08-28T18:09:40.0860651Z   File "/var/lib/jenkins/.local/lib/python3.7/site-packages/torchdynamo/variables/functions.py", line 31, in wrap_bound_arg
2022-08-28T18:09:40.0861268Z     assert isinstance(val, VariableTracker), typestr(val)
2022-08-28T18:09:40.0862427Z AssertionError: memory_format
2022-08-28T18:09:40.0862666Z 
2022-08-28T18:09:40.0862790Z from user code:
2022-08-28T18:09:40.0863142Z    File "test_nn.py", line 13726, in test_Dropout1d
2022-08-28T18:09:40.0865885Z     self._test_dropout(nn.Dropout1d, device, input)
2022-08-28T18:09:40.0866139Z 
2022-08-28T18:09:40.0866370Z Set torchdynamo.config.verbose=True for more information
2022-08-28T18:09:40.0866728Z ==========
```

Reply:
cc @ezyang what should we do with these failures?

Reply:
cc @anijain2305 

You can unblock by using `@skipIfTorchDynamo`

Reply:
This PR is not blocked by the same.